src/mlpack/methods/ann/layer/binary_rbm.hpp

```diff
@@ -0,0 +1,130 @@
+/**
+ * @file binary_rbm.hpp
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_BINARY_RBM_HPP
+#define MLPACK_METHODS_ANN_LAYER_BINARY_RBM_HPP
+
+#include "layer_types.hpp"
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/activation_functions/logistic_function.hpp>
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+
+
+namespace mlpack{
+namespace ann{
+
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class VisibleLayer
+{
+ public:
+  /* The visible layer of the rbm
+   * network.
+   *
+   * @param: inSize: num of visible neurons
+   * @param: outSize: num of hidden neurons
+   */
+  VisibleLayer(const size_t inSize, const size_t outSize);
+
+  // Reset the variables
+  void Reset();
+
+  /**
+    * Calculate the acivations and send to the hidden layer.
+    * input if of the format datapoint + bias(other layer)
+    *
+    * @param input Input data used for evaluating the specified function.
+    * @param output Resulting output activation.
+    */
+  template<typename eT>
+  void Forward(arma::Mat<eT>&& input, arma::Mat<eT>&& output);
+
+  /**
+   * Sample the output given the input parameters
+   * The sample are obtained from a distribution 
+   * specified by the sampler. 
+   *
+   * @param input the input parameters for 
+   * @param output samples from the parameters
+   */
+  template<typename eT>
+  void Sample(arma::Mat<eT>&& input, arma::Mat<eT>&& output);
```

Comment:
If we look at the deeplearning.net example. Forward function actually computes the activation form the sigmoid layer and sample actually uses the activations to sample visible/ hidden layer. I think we need both